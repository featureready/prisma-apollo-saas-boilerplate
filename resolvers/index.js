import { GraphQLDateTime } from 'graphql-iso-date';

import userResolvers from './user';
import teamResolvers from './team';

const customScalarResolver = {
  DateTime: GraphQLDateTime
};

export default [customScalarResolver, userResolvers, teamResolvers];
