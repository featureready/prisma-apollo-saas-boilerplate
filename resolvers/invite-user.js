import _ from 'lodash';
import sendEmail from '../services/sendgrid';

export default {
  Mutation: {
    async InviteUser(parent, { input }, { user, prisma }, info) {
      const { email, firstName, lastName } = input;
      const team = await prisma.user({ id: user.id }).team();
      const invite = await prisma.createInvite({
        email,
        firstName,
        lastName,
        user: {
          connect: {
            id: user.id
          }
        },
        team: {
          connect: {
            id: team.id
          }
        }
      });

      sendEmail({
        template_id: process.env.INVITE_USER_TEMPLATE_ID,
        to: email,
        from: process.env.SUPPORT_EMAIL_ADDRESS,
        dynamic_template_data: {
          invitee: `${firstName} ${lastName}`,
          inviter: user.email,
          signUpLink: `https://app.zapcms.com/signup?inviteId=${invite.id}`
        }
      });

      return invite;
    }
  }
};
