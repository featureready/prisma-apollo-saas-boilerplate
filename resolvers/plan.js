import _ from 'lodash';
import { listAllPlans } from '../services/stripe';
import { toCamelCase } from '../helpers/arrayUtils';

export default {
  Query: {
    async Plans(parent, args, context, info) {
      const plans = await listAllPlans();

      return toCamelCase(plans);
    }
  }
};
