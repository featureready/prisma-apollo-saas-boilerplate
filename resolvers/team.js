import { analytics } from '../services/segment';

export default {
  Mutation: {
    async createTeam(parent, { name }, { user, models }, info) {
      const team = await models.team.create({
        name
      });
      await user.update({ teamId: team.id });

      return team;
    },
    async updateTeam(parent, { name }, { user, models }, info) {
      const dbUser = await models.user.findOne({
        where: { id: user.id },
        include: [{ model: models.team }]
      });

      if (dbUser.team) {
        return dbUser.team.update({ name });
      }

      throw new Error('Unable to update team - please create a team first.');
    }
  }
};
