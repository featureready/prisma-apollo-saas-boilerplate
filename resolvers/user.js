import bcrypt from 'bcrypt';
import jsonwebtoken from 'jsonwebtoken';
import uuidv4 from 'uuid/v4';
import { UserInputError, AuthenticationError } from 'apollo-server';
import sendEmail from '../services/sendgrid';
import {
  createCustomer,
  createCard,
  createSubscription,
  listAllInvoices
} from '../services/stripe';
import { toCamelCase } from '../helpers/arrayUtils';

export default {
  Query: {
    async me(parent, args, { user, prisma }, info) {
      return prisma.user({ id: user.id });
    },
    async paymentHistory(parent, args, context, info) {
      const { stripeCustomerId: customerId } = context.user;
      const invoices = await listAllInvoices({
        customerId
      });

      return toCamelCase(invoices);
    }
  },
  Mutation: {
    async signup(parent, { firstName, lastName, email, password }, { prisma }, info) {
      if (await prisma.user({ email })) {
        throw new UserInputError('Email is already taken');
      } else {
        const user = await prisma.createUser({
          firstName,
          lastName,
          email,
          password: hashedPassword(password),
          stripeCustomerId: await createCustomer({ email })
        });

        return {
          user,
          jwt: getJwt({
            id: user.id,
            email: user.email
          })
        };
      }
    },
    async login(parent, { email, password }, { prisma }, info) {
      const user = await prisma.user({ email });

      if (user && bcrypt.compareSync(password, user.password)) {
        return {
          user,
          jwt: getJwt({
            id: user.id,
            email: user.email
          })
        };
      }

      throw new AuthenticationError('Please check your credentials and try again.');
    },
    async forgotPassword(parent, { email }, { prisma }, info) {
      const user = await prisma.user({ email });

      if (user) {
        const { email: userEmail } = user;
        const token = uuidv4();

        await prisma.updateUser({
          where: { email },
          data: { resetPasswordToken: token }
        });

        sendEmail({
          template_id: process.env.FORGOT_PASSWORD_TEMPLATE_ID,
          to: userEmail,
          from: process.env.SUPPORT_EMAIL_ADDRESS,
          dynamic_template_data: {
            email: userEmail,
            resetPasswordLink: `https://app.zapcms.com/reset-password?token=${token}`
          }
        });
      }

      return {
        message: 'A link to reset your password will be sent to your registered email.'
      };
    },
    async resetPassword(parent, { password, token }, { prisma }, info) {
      const dbUser = await prisma.user({ resetPasswordToken: token });

      if (dbUser) {
        await prisma.updateUser({
          where: { resetPasswordToken: token },
          data: {
            password: hashedPassword(password),
            resetPasswordToken: null
          }
        });

        return {
          message: 'Password updated successfully.'
        };
      }

      throw new AuthenticationError('Password reset token is invalid.');
    },
    async updateUser(parent, { firstName, lastName, email, password }, { user, prisma }, info) {
      if (password) {
        return prisma.updateUser({
          where: { id: user.id },
          data: {
            firstName,
            lastName,
            email,
            password: hashedPassword(password)
          }
        });
      }

      return prisma.updateUser({
        where: { id: user.id },
        data: {
          firstName,
          lastName,
          email
        }
      });
    },
    async deleteUser(
      parent,
      args,
      {
        user: { id },
        prisma
      },
      info
    ) {
      return prisma.deleteUser({ id });
    },
    async addCreditCard(parent, args, context, info) {
      const { stripeCustomerId: customerId } = context.user;
      const { token } = args;

      await createCard({
        token,
        customerId
      });

      return {
        message: 'Successfully updated billing information.'
      };
    },
    async subscribePlan(parent, args, context, info) {
      const { planId } = args;
      const { stripeCustomerId: customerId } = context.user;

      await createSubscription({
        planId,
        customerId
      });

      return {
        message: 'Successfully subscribed to plan.'
      };
    }
  }
};

const hashedPassword = password => bcrypt.hashSync(password, 10);
const getJwt = ({ id, email }) => jsonwebtoken.sign({ id, email }, process.env.JWT_SECRET);
