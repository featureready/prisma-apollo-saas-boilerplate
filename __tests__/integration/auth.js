import { createTestClient } from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import typeDefs from '../../schema';
import resolvers from '../../resolvers';
import schemaDirectives from '../../directives';
import userFactory from '../../factories/user';
import {
  ME,
  PAYMENT_HISTORY,
  SIGNUP,
  LOGIN,
  UPDATE_USER,
  FORGOT_PASSWORD,
  DELETE_USER,
  ADD_CREDIT_CARD,
  SUBSCRIBE_PLAN
} from '../../queries/user';

let client;

beforeEach(() => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    schemaDirectives,
    context: () => ({
      prisma: {
        user: ({ id }) => userFactory({ id })
      }
    })
  });

  client = createTestClient(server);
});

it('shows authentication error when fetching Me', async () => {
  const res = await client.query({
    query: ME
  });

  expect(res).toMatchSnapshot();
});
