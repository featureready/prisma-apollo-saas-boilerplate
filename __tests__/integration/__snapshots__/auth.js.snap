// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows authentication error when fetching Me 1`] = `
Object {
  "data": Object {
    "me": null,
  },
  "errors": Array [
    [GraphQLError: You must be authenticated to perform this action],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
